### PR TITLE
Add support for existing S3 buckets

### LINKED ISSUE #4241: Introduce a functionality which attach events to an existing resource
# This is a Feature Proposal

## Description
For feature proposals:
* What is the use case that should be solved. The more detail you describe this in the easier it is to understand for us.

  `S3` and `cognitoUserPool` events can't attach an existing resource. I think the main reason is that those can't specify an external resource within CF generated by Serverless.

However, I think many users hope that. 
I think there are many cases that data sources like S3, DynamoDB or  Cognito userpool are preferred to separate a stack from a one which includes Lambda.
 If those are built in a same service, you can't remove and recreate anytime since data sources have a state.

* If there is additional config how would it look

Similar or dependent issues:
* #4237, #4207, #2154

### PR DESCRIPTION
**Pull request by pmuens**
This PR adds support for existing S3 buckets via a custom resource.

## What did you implement:

Refs #4241

## How did you implement it:

Checks the `existing` config property of the `s3` event and swaps out the S3 CloudFormation resource with a custom resource the Serverless Framework provides. The custom resource basically does raw SDK calls to update the existing S3 bucket. This way everything is encapsulated via CloudFormation and the stack won't get out of sync if we implement it via raw SDK calls.

## How can we verify it:

```yaml
service: test

provider:
  name: aws
  runtime: nodejs10.x

functions:
  test:
    handler: handler.hello
    events:
      - s3:
          bucket: existing-s3-bucket
          event: s3:ObjectCreated:*
          rules:
            - prefix: uploads
            - suffix: .jpg
          existing: true
```

## Todos:

- [x] Write tests
- [x] Write documentation
- [x] Fix linting errors
- [x] Make sure code coverage hasn't dropped
- [x] Provide verification config / commands / resources
- [x] Enable "Allow edits from maintainers" for this PR
- [x] Update the messages below

***Is this ready for review?:*** YES
***Is it a breaking change?:*** NO

### CODE REVIEW

**Comment by raeesbhatti** [comment]
Looking forward to this. Thanks 👍🏼

**Comment by garethmcc** [comment]
I know many many community members who are looking forward to this, myself included. Nice one!!!!

File: lib/plugins/aws/customResources/index.js [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+const createZipFile = require('../../../utils/fs/createZipFile');
+
+function addCustomResourceToService(resourceName) {
+  let FunctionName;
+  let Handler;
+
+  let customResourceFunctionLogicalId;
+  const srcDirPath = path.join(__dirname, 'resources');
+  // TODO: not sure if this is deterministic given that we have packaging options
+  // on the service and function level
```
Comment (pmuens) [inline lib/plugins/aws/customResources/index.js]: @medikoo do you know the answer to this question?

File: lib/utils/fs/createZipFile.js [hunk 2]
```diff
@@ -0,0 +1,45 @@
+'use strict';
+
+const fs = require('fs');
+const path = require('path');
+const globby = require('globby');
+const archiver = require('archiver');
+const BbPromise = require('bluebird');
+
+function createZipFile(srcDirPath, outputFilePath) {
+  const files = globby
```
Comment (medikoo) [inline lib/utils/fs/createZipFile.js]: If we're after all files, wouldn't it be more to the point just use some recursive readdir? `globby` is more for pattern matching, so seems a bit overkill for this case

(assuming we have some util handy for that)

File: lib/classes/Utils.js [hunk 3]
```diff
@@ -133,6 +134,10 @@ class Utils {
     return servicePath;
   }
 
+  createZipFile(srcDirPath, outputFilePath) {
+    return createZipFile(srcDirPath, outputFilePath);
+  }
```
[the remark this answers is not in the document]
Comment (pmuens) [reply lib/classes/Utils.js]: Yes, I just added it to follow our pattern here. Let me remove it for now to get rid of yet another maintenance burden.

We can bring it back if we need it.

File: lib/plugins/aws/customResources/index.js (hunk 1 above)
Comment (pmuens) [reply lib/plugins/aws/customResources/index.js]: @medikoo you were right. I just tested it with a custom artifact and it works just fine :+1:

File: lib/plugins/aws/customResources/resources/s3/handler.js [hunk 4]
```diff
@@ -0,0 +1,85 @@
...
+
+  return updateConfiguration({
+    lambdaArn,
+    region: Region,
+    functionName: FunctionName,
+    bucketName: BucketName,
+    bucketConfig: BucketConfig,
+  })
+    .then(() => response(event, context, 'SUCCESS'))
+    .catch(error => response(event, context, 'FAILED', {}, error));
```
Comment (medikoo) [inline lib/plugins/aws/customResources/resources/s3/handler.js]: What about generalizing a response handling to util? I think it'll reduce a significant amount of noise from endpoints here.

Do you plan to cover it with when adding support  for existing Cognito user pools?

File: lib/plugins/aws/package/compile/events/s3/index.js [hunk 5]
```diff
@@ -1,26 +1,37 @@
...
 
-  compileS3Events() {
+  newS3Buckets() {
     const bucketsLambdaConfigurations = {};
     const s3EnabledFunctions = [];
     this.serverless.service.getAllFunctions().forEach(functionName => {
       const functionObj = this.serverless.service.getFunction(functionName);
 
       if (functionObj.events) {
+        // eslint-disable-next-line consistent-return
```
Comment (medikoo) [inline lib/plugins/aws/package/compile/events/s3/index.js]: I don't think this disable is needed. At below line we should just do `return;` instead of `return null`, and there shold be no error reported by linter.

File: lib/plugins/aws/package/compile/events/s3/index.js [hunk 6]
```diff
@@ -1,26 +1,37 @@
...
         functionObj.events.forEach(event => {
           if (event.s3) {
+            // return immediately if it's an existing S3 event since we treat them differently
+            if (_.isObject(event.s3) && event.s3.existing) return null;
```
Comment (medikoo) [inline lib/plugins/aws/package/compile/events/s3/index.js]: nit: `_isObject` check is not needed.

File: tests/integration-all/s3/service/core.js [hunk 7]
```diff
@@ -0,0 +1,30 @@
+'use strict';
+
+function minimal(event, context, callback) {
+  const response = { message: 'Hello from S3! - (minimal)', event };
+  process.stdout.write(event.Records[0].eventSource);
+  process.stdout.write(event.Records[0].eventName);
```
Comment (medikoo) [inline tests/integration-all/s3/service/core.js]: Do those writes pay any role? Also there's no whitespace between writes

File: tests/integration-all/s3/tests.js [hunk 8]
```diff
@@ -0,0 +1,104 @@
...
+    console.info(`Deleting S3 bucket "${bucketExistingSetup}"...`);
+    return deleteBucket(bucketExistingSetup);
+  });
+
+  describe('Minimal Setup', () => {
+    const expectedMessage = 'Hello from S3! - (minimal)';
+
+    it('should invoke function when an object is created', () => {
+      return createAndRemoveInBucket(bucketMinimalSetup)
+        .then(() => delayPromise(delay))
```
Comment (medikoo) [inline tests/integration-all/s3/tests.js]: Such approach is prone to false positives (in some corner case scenario, where logs will appear a bit later it'll fail test), and also it unnecessary prolongs testing (e.g. if usually logs are visible after 2s, we always wait 15s)

IMO best approach is to check  logs on interval (e.g. every 1 second), with some max time setting (30 seconds?). We can prepare some neat util that will help us handle that cleanly.

File: lib/plugins/aws/customResources/resources/s3/handler.js (hunk 4 above)
Comment (pmuens) [reply lib/plugins/aws/customResources/resources/s3/handler.js]: I like the idea but I'm not sure if I fully understand how this handling util would look like.

Can you provide any pointers?

Comment (medikoo) [reply lib/plugins/aws/customResources/resources/s3/handler.js]: in `utils.js` we can expose handler which will work as handler decorator, e.g.

```javascript
modul.exports.awsHandler = main => {
  return (event, context, callback) => {
    Promise.resolve(main()).then(
      result => response(event, context, 'SUCCESS', result),
      error => response(event, context, 'FAILED', {}, error))
    ).then(result => callback(null, result), callback);
  }
}
```

And then in `handler.js` we can do:

```javascript
const { awsHandler } = require('./utils');
...
module.exports: = {
  handler: awsHandler(handler)
}
```

File: lib/plugins/aws/package/compile/events/s3/index.js (hunk 5 above)
Comment (pmuens) [reply lib/plugins/aws/package/compile/events/s3/index.js]: I can remove this line but we still need to return `null` (not `return;`). Otherwise we get the same consistent return violation error.

File: tests/integration-all/s3/service/core.js (hunk 7 above)
Comment (pmuens) [reply tests/integration-all/s3/service/core.js]: Yes, that's by design. We have to write so that the output shows up in the CloudWatch logs. We could also use `console.log` but I thought that using `process.stdout.write` would be more expressive.

File: lib/plugins/aws/package/compile/events/s3/index.js (hunk 5 above)
Comment (medikoo) [reply lib/plugins/aws/package/compile/events/s3/index.js]: Ok, good, my point was, that it's best if we rely on `eslint-disable-next-line` only where there's absolutely no way to go without it (and we want to keep rule on, for some good reasons).

If `return null` is the way, then it's totally fine :)

File: tests/integration-all/s3/service/core.js (hunk 7 above)
Comment (medikoo) [reply tests/integration-all/s3/service/core.js]: Ok, but is it not error-prone, that we output `eventSource` and `eventName` without an interim whitespace? As configured above, in log it'll appear as e.g. `eventSourceeventName`

File: lib/plugins/aws/customResources/resources/s3/handler.js (hunk 4 above)
Comment (pmuens) [reply lib/plugins/aws/customResources/resources/s3/handler.js]: Ah, that makes sense :+1:

Just implemented this and will ship it with the next PR update.

File: tests/integration-all/s3/tests.js (hunk 8 above)
Comment (pmuens) [reply tests/integration-all/s3/tests.js]: I like that. Will ship a version of this in the upcoming changes.

File: tests/integration-all/s3/service/utils.js [hunk 9]
```diff
@@ -0,0 +1,20 @@
+'use strict';
+
+function getMarkers(functionName) {
+  return {
+    start: `--- START ${functionName} ---`,
+    end: `--- END ${functionName} ---`,
```
Comment (medikoo) [inline tests/integration-all/s3/service/utils.js]: Wouldn't it be better if there's new line after START and before END?

File: tests/integration-all/s3/service/utils.js [hunk 10]
```diff
@@ -0,0 +1,20 @@
...
+  };
+}
+
+function log(functionName, message) {
+  const markers = getMarkers(functionName);
+  process.stdout.write(markers.start);
+  process.stdout.write(message);
+  process.stdout.write(markers.end);
```
Comment (medikoo) [inline tests/integration-all/s3/service/utils.js]: If it's supposed to be run in AWS lambda, then using `process.stdout.write` over `console.log` has implication of log being written out of convention (no request id and no timestamp)

Is that intentional?

Comment (pmuens) [reply tests/integration-all/s3/service/utils.js]: AFAIK it won't impact the correctness here. I'll update it regardless since it seems to be more intuitive to use `console.log` in that case.

File: tests/integration-all/s3/service/utils.js (hunk 9 above)
Comment (pmuens) [reply tests/integration-all/s3/service/utils.js]: Will update it.

**Comment by shawn271828** [comment]
Great job!

I came across searching for this feature and have discovered a tiny issue:

Resources in policy statement have hard-coded `xxx:aws:xxx`.
This seems to be an issue in some region like China when it is `xxx:aws-cn:xxx`

**Comment by pmuens** [comment]
Thanks for the feedback @shawn271828 👍

Can you open up an issue for this so that we can track and address it? Thanks!

**Comment by coffbr01** [comment]
@pmuens, thank you for this patch. This change saved me a ton of development time. You have achieved hero status in my book.

**Comment by pmuens** [comment]
> @pmuens, thank you for this patch. This change saved me a ton of development time. You have achieved hero status in my book.

Hah. Thanks a ton for that feedback @coffbr01 🙏 💯 :+1:

**Comment by pavelloz** [comment]
Holy moly this is awesome
